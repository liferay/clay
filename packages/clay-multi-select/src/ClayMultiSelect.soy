{namespace ClayMultiSelect}

/**
 * This renders the component's whole content.
 */
{template .render}
	{@param dataSource: string|[]|list<?>}
	{@param helpText: string}
	{@param spritemap: string}
	{@param? _handleButtonClicked: any}
	{@param? _handleCloseButtonClick: any}
	{@param? _handleDataChange: any}
	{@param? _handleDropdownItemClick: any}
	{@param? _handleFilteredItemsChange: any}
	{@param? _handleOnInput: any}
	{@param? _handleOnKeydown: any}
	{@param? _removeFocusedItem: any}
	{@param? contentRenderer: string}
	{@param? debounceTime: number}
	{@param? elementClasses: string}
	{@param? enableAutocomplete: bool}
	{@param? filteredItems: list<?>}
	{@param? id: string}
	{@param? initialData: []|list<?>}
	{@param? inputName: string}
	{@param? label: string}
	{@param? labelLocator: any}
	{@param? requestOptions: [
		method: string,
		mode: string,
		cache: string,
		credentials: string,
		headers: [],
		redirect: string,
		referrer: string,
		body: []
	]}
	{@param? pollingInterval: number}
	{@param? requestRetries: number}
	{@param? requestTimeout: number}
	{@param? selectedItems: list<?>}

	{let $attributes kind="attributes"}
		class="from-group
			{if $elementClasses}
				{sp}{$elementClasses}
			{/if}
		"

		{if $id}
			id="{$id}"
		{/if}
	{/let}

	<div {$attributes}>
		{if $label}
			<label>{$label}</label>
		{/if}

		{call .content}
			{param _handleButtonClicked: $_handleButtonClicked /}
			{param _handleCloseButtonClick: $_handleCloseButtonClick /}
			{param _handleDataChange: $_handleDataChange /}
			{param _handleDropdownItemClick: $_handleDropdownItemClick /}
			{param _handleFilteredItemsChange: $_handleFilteredItemsChange /}
			{param _handleOnInput: $_handleOnInput /}
			{param _handleOnKeydown: $_handleOnKeydown /}
			{param _removeFocusedItem: $_removeFocusedItem /}
			{param contentRenderer: $contentRenderer ?: ''/}
			{param dataSource: $dataSource /}
			{param debounceTime: $debounceTime /}
			{param enableAutocomplete: $enableAutocomplete /}
			{param filteredItems: $filteredItems /}
			{param helpText: $helpText /}
			{param initialData: $initialData /}
			{param inputName: $inputName /}
			{param labelLocator: $labelLocator /}
			{param requestOptions: $requestOptions /}
			{param pollingInterval: $pollingInterval /}
			{param requestRetries: $requestRetries /}
			{param requestTimeout: $requestTimeout /}
			{param selectedItems: $selectedItems /}
			{param spritemap: $spritemap /}
		{/call}
	</div>
{/template}

/**
 * This render the content from the multi select.
 */
{template .content}
	{@param dataSource: string|[]|list<?>}
	{@param helpText: string}
	{@param spritemap: string}
	{@param? _handleButtonClicked: any}
	{@param? _handleCloseButtonClick: any}
	{@param? _handleDataChange: any}
	{@param? _handleDropdownItemClick: any}
	{@param? _handleFilteredItemsChange: any}
	{@param? _handleOnInput: any}
	{@param? _handleOnKeydown: any}
	{@param? _removeFocusedItem: any}
	{@param? contentRenderer: string}
	{@param? debounceTime: number}
	{@param? enableAutocomplete: bool}
	{@param? filteredItems: list<?>}
	{@param? initialData: []|list<?>}
	{@param? inputName: string}
	{@param? labelLocator: any}
	{@param? requestOptions: [
		method: string,
		mode: string,
		cache: string,
		credentials: string,
		headers: [],
		redirect: string,
		referrer: string,
		body: []
	]}
	{@param? pollingInterval: number}
	{@param? requestRetries: number}
	{@param? requestTimeout: number}
	{@param? selectedItems: list<?>}

	<div class="input-group input-group-stacked-sm-down">
		<div class="input-group-item">
			{call ClayAutocomplete.render}
				{param contentRenderer: $contentRenderer /}
				{param dataSource: $dataSource /}
				{param debounceTime: $debounceTime /}
				{param elementClasses: 'form-control form-control-tag-group' /}
				{param enableAutocomplete: $enableAutocomplete /}
				{param events: [
					'dataChange': $_handleDataChange,
					'filteredItemsChanged': $_handleFilteredItemsChange,
					'inputChange': $_handleOnInput,
					'inputOnBlur': $_removeFocusedItem,
					'inputOnFocus': $_removeFocusedItem,
					'inputOnKeydown': $_handleOnKeydown,
					'itemSelected': $_handleDropdownItemClick
				] /}
				{param extractData: $labelLocator /}
				{param filteredItems: $filteredItems /}
				{param initialData: $initialData /}
				{param inputElementClasses: 'form-control-inset' /}
				{param ref: 'autocomplete' /}
				{param requestOptions: $requestOptions /}
				{param pollingInterval: $pollingInterval /}
				{param requestRetries: $requestRetries /}
				{param requestTimeout: $requestTimeout /}
				{param unstable_content kind="html"}
					{if $selectedItems}
						{call .items}
							{param _handleCloseButtonClick: $_handleCloseButtonClick /}
							{param contentRenderer: $contentRenderer /}
							{param inputName: $inputName /}
							{param selectedItems: $selectedItems /}
							{param spritemap: $spritemap /}
						{/call}
					{/if}
				{/param}
			{/call}

			{let $isVisible: $filteredItems and length($filteredItems) > 0 ? true : false /}

			{if not $isVisible}
				<div class="form-feedback-group">
					<div class="form-text">
						{$helpText}
					</div>
				</div>
			{/if}
		</div>
		<div class="input-group-item input-group-item-shrink">
			{call ClayButton.render}
				{param events: ['click': $_handleButtonClicked] /}
				{param label kind="text"}
					{msg desc="Select items"}
						select
					{/msg}
				{/param}

				{param ref: 'button' /}
				{param style: 'secondary' /}
			{/call}
		</div>
	</div>
{/template}

/**
 * This render the selected items.
 */
{template .items}
	{@param selectedItems: list<?>}
	{@param spritemap: string}
	{@param? _handleCloseButtonClick: any}
	{@param? contentRenderer: string}
	{@param? inputName: string}

	{foreach $item in $selectedItems}
		{if $inputName}
			{let $inputAttributes kind="attributes"}
				{if $inputName}
					name="{$inputName}"
				{/if}

				type="hidden"
				value="{$item.value}"
			{/let}

			<input {$inputAttributes}>
		{/if}

		{delcall ClayMultiSelect.Item variant="$contentRenderer" allowemptydefault="true"}
			{param _handleCloseButtonClick: $_handleCloseButtonClick /}
			{param data: $item /}
			{param index: index($item) /}
			{param spritemap: $spritemap /}
		{/delcall}
	{/foreach}
{/template}

/**
 * This render the item.
 * Extension point: Create a new variation above ClayMultiSelect.Item
 * so you can create your own markup.
 */
{deltemplate ClayMultiSelect.Item}
	{@param data: ?}
	{@param index: int}
	{@param spritemap: string}
	{@param? _handleCloseButtonClick: any}

	{call ClayLabel.render}
		{param closeable: true /}
		{param data: ['tag': $index] /}
		{param events: ['close': $_handleCloseButtonClick] /}
		{param id: 'item-tag' /}
		{param label: $data.label /}
		{param ref: 'item' + $index /}
		{param spritemap: $spritemap /}
		{param tabIndex: '0' /}
	{/call}
{/deltemplate}
