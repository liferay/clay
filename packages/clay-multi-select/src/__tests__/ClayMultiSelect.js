import {triggerEvent} from 'metal-dom';
import ClayMultiSelect from '../ClayMultiSelect';

let component;
let dataSource = [
	'Bread',
	'Ammonia cookie',
	'Cuisine of Antebellum America',
	'Apple butter',
	'Apple sauce',
	'Baked potato',
	'Barbecue',
	'Bear claw',
	'Beef Manhattan',
	'Blue cheese dressing',
	'Blue-plate special',
	'Bookbinder soup',
	'Breakfast burrito',
	'Brunswick stew',
	'Buffalo burger',
	'Buffalo wing',
	'Bull roast',
	'Burnt ends',
	'Butter cookie',
];
let spritemap = 'icons.svg';
let helpText = 'You can use a comma to enter tags';

describe('ClayMultiSelect', function() {
	beforeEach(() => {
		jest.useFakeTimers();
	});

	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('should render the default markup', () => {
		component = new ClayMultiSelect({
			dataSource,
			helpText,
			spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayMultiSelect with classes', () => {
		component = new ClayMultiSelect({
			dataSource,
			elementClasses: 'my-custom-class',
			helpText,
			spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayMultiSelect with id', () => {
		component = new ClayMultiSelect({
			dataSource,
			helpText,
			id: 'myId',
			spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayMultiSelect with label', () => {
		component = new ClayMultiSelect({
			dataSource,
			helpText,
			inputName: 'foo',
			label: 'Foo',
			spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayMultiSelect with selected items', () => {
		component = new ClayMultiSelect({
			dataSource,
			helpText,
			selectedItems: [
				{
					label: 'Foo',
					value: 'foo',
				},
				{
					label: 'Bar',
					value: 'bar',
				},
			],
			spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayMultiSelect with input name from selected items', () => {
		component = new ClayMultiSelect({
			dataSource,
			helpText,
			inputName: 'foo',
			spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	describe('Lifecycle', () => {
		it('should emit an event when the button is clicked', () => {
			component = new ClayMultiSelect({
				dataSource,
				helpText,
				spritemap,
			});

			const spy = jest.spyOn(component, 'emit');
			const {button} = component.refs;

			button.element.click();

			expect(spy).toHaveBeenCalled();
			expect(spy).toHaveBeenCalledWith(
				expect.objectContaining({
					name: 'buttonClicked',
					originalEvent: expect.any(Object),
				})
			);
		});

		it('should emit an event when the dropdown item is selected', () => {
			component = new ClayMultiSelect({
				dataSource,
				helpText,
				selectedItems: [
					{
						label: 'Foo',
						value: 'foo',
					},
					{
						label: 'Bar',
						value: 'bar',
					},
				],
				spritemap,
			});

			const spy = jest.spyOn(component, 'emit');
			const {input} = component.refs.autocomplete.refs;

			input.value = 'bre';
			triggerEvent(input, 'input', {data: 'a'});
			triggerEvent(input, 'keydown', {key: 'ArrowDown'});

			jest.runAllTimers();

			const {
				dropdown,
			} = component.refs.autocomplete.refs.dataProvider.refs;

			dropdown.querySelector('a[data-dropdown-item-index]').click();

			expect(spy).toHaveBeenCalled();
			expect(spy).toHaveBeenCalledWith(
				expect.objectContaining({
					data: expect.any(Object),
					name: 'itemSelected',
					originalEvent: expect.any(Object),
				})
			);
		});

		it('should emit an event when the input changed', () => {
			component = new ClayMultiSelect({
				dataSource,
				helpText,
				spritemap,
			});

			const spy = jest.spyOn(component, 'emit');
			const {input} = component.refs.autocomplete.refs;

			input.value = 'foo';
			triggerEvent(input, 'input', {data: 'o'});

			jest.runAllTimers();

			expect(spy).toHaveBeenCalled();
			expect(spy).toHaveBeenCalledWith(
				expect.objectContaining({
					data: {
						value: 'foo',
					},
					name: 'inputChange',
					originalEvent: expect.any(Object),
				})
			);
		});

		describe('itemAdded', () => {
			it('should emit an event when press enter on the input', () => {
				component = new ClayMultiSelect({
					dataSource,
					helpText,
					spritemap,
				});

				const spy = jest.spyOn(component, 'emit');
				const {input} = component.refs.autocomplete.refs;

				input.value = 'foo';
				triggerEvent(input, 'keydown', {key: 'Enter'});

				jest.runAllTimers();

				expect(spy).toHaveBeenCalled();
				expect(spy).toHaveBeenCalledWith(
					expect.objectContaining({
						data: {
							item: {
								label: 'foo',
								value: 'foo',
							},
						},
						name: 'itemAdded',
						originalEvent: expect.any(Object),
					})
				);
			});

			it('should emit an event when press comma on the input', () => {
				component = new ClayMultiSelect({
					dataSource,
					helpText,
					spritemap,
				});

				const spy = jest.spyOn(component, 'emit');
				const {input} = component.refs.autocomplete.refs;

				input.value = 'bar';
				triggerEvent(input, 'input', {data: ','});

				jest.runAllTimers();

				expect(spy).toHaveBeenCalled();
				expect(spy).toHaveBeenCalledWith(
					expect.objectContaining({
						data: {
							item: {
								label: 'bar',
								value: 'bar',
							},
						},
						name: 'itemAdded',
						originalEvent: expect.any(Object),
					})
				);
			});

			it('should not emit an event when the input value exists on the selected items', () => {
				component = new ClayMultiSelect({
					dataSource,
					helpText,
					selectedItems: [
						{
							label: 'foo',
							value: 'foo',
						},
					],
					spritemap,
				});

				const spy = jest.spyOn(component, 'emit');
				const {input} = component.refs.autocomplete.refs;

				input.value = 'foo';
				triggerEvent(input, 'keydown', {key: 'Enter'});

				jest.runAllTimers();

				expect(spy).not.toHaveBeenCalled();
				expect(spy).not.toHaveBeenCalledWith(
					expect.objectContaining({
						data: {
							item: {
								label: 'foo',
								value: 'foo',
							},
						},
						name: 'itemAdded',
						originalEvent: expect.any(Object),
					})
				);
				expect(component).toMatchSnapshot();
			});
		});

		describe('itemRemoved', () => {
			it('should emit an event when the button close from item is clicked', () => {
				component = new ClayMultiSelect({
					dataSource,
					helpText,
					selectedItems: [
						{
							label: 'Foo',
							value: 'foo',
						},
						{
							label: 'Bar',
							value: 'bar',
						},
					],
					spritemap,
				});

				const spy = jest.spyOn(component, 'emit');
				const {autocomplete} = component.refs;

				autocomplete.refs.item1.refs.closeButton.element.click();

				expect(spy).toHaveBeenCalled();
				expect(spy).toHaveBeenCalledWith(
					expect.objectContaining({
						data: {
							item: {
								label: 'Bar',
								value: 'bar',
							},
						},
						name: 'itemRemoved',
						originalEvent: expect.any(Object),
					})
				);
			});

			it('should emit an event when the item is focused and press backspace', () => {
				component = new ClayMultiSelect({
					dataSource,
					helpText,
					selectedItems: [
						{
							label: 'Foo',
							value: 'foo',
						},
						{
							label: 'Bar',
							value: 'bar',
						},
					],
					spritemap,
				});

				const spy = jest.spyOn(component, 'emit');
				const {autocomplete} = component.refs;

				triggerEvent(autocomplete.refs.input, 'keydown', {
					key: 'Backspace',
				});

				jest.runAllTimers();

				const items = autocomplete.element.querySelectorAll(
					'span[id="item-tag"]'
				);

				expect(items[1]).toEqual(document.activeElement);

				triggerEvent(autocomplete.element, 'keydown', {
					key: 'Backspace',
				});

				expect(spy).toHaveBeenCalled();
				expect(spy).toHaveBeenCalledWith(
					expect.objectContaining({
						data: {
							item: {
								label: 'Bar',
								value: 'bar',
							},
						},
						name: 'itemRemoved',
						originalEvent: expect.any(Object),
					})
				);
			});

			it('should emit an event when the item is focused and enter backspace', () => {
				component = new ClayMultiSelect({
					dataSource,
					helpText,
					selectedItems: [
						{
							label: 'Foo',
							value: 'foo',
						},
						{
							label: 'Bar',
							value: 'bar',
						},
					],
					spritemap,
				});

				const spy = jest.spyOn(component, 'emit');
				const {autocomplete} = component.refs;

				triggerEvent(autocomplete.refs.input, 'keydown', {
					key: 'Backspace',
				});

				jest.runAllTimers();

				const items = autocomplete.element.querySelectorAll(
					'span[id="item-tag"]'
				);

				expect(items[1]).toEqual(document.activeElement);

				triggerEvent(autocomplete.element, 'keydown', {key: 'Enter'});

				expect(spy).toHaveBeenCalled();
				expect(spy).toHaveBeenCalledWith(
					expect.objectContaining({
						data: {
							item: {
								label: 'Bar',
								value: 'bar',
							},
						},
						name: 'itemRemoved',
						originalEvent: expect.any(Object),
					})
				);
			});
		});
	});

	describe('with keyboard interaction', () => {
		it('should focus on the left item when press the arrow left', () => {
			component = new ClayMultiSelect({
				dataSource,
				helpText,
				selectedItems: [
					{
						label: 'Foo',
						value: 'foo',
					},
					{
						label: 'Bar',
						value: 'bar',
					},
				],
				spritemap,
			});

			const {autocomplete} = component.refs;

			triggerEvent(autocomplete.refs.input, 'keydown', {
				key: 'Backspace',
			});

			jest.runAllTimers();

			const items = autocomplete.element.querySelectorAll(
				'span[id="item-tag"]'
			);

			expect(items[1]).toEqual(document.activeElement);

			triggerEvent(autocomplete.element, 'keydown', {key: 'ArrowLeft'});
			expect(items[0]).toEqual(document.activeElement);
		});

		it('should focus on the right item when press the arrow right', () => {
			component = new ClayMultiSelect({
				dataSource,
				helpText,
				selectedItems: [
					{
						label: 'Foo',
						value: 'foo',
					},
					{
						label: 'Bar',
						value: 'bar',
					},
				],
				spritemap,
			});

			const {autocomplete} = component.refs;

			triggerEvent(autocomplete.refs.input, 'keydown', {
				key: 'Backspace',
			});

			jest.runAllTimers();

			const items = autocomplete.element.querySelectorAll(
				'span[id="item-tag"]'
			);

			expect(items[1]).toEqual(document.activeElement);

			triggerEvent(autocomplete.element, 'keydown', {key: 'ArrowLeft'});
			expect(items[0]).toEqual(document.activeElement);

			triggerEvent(autocomplete.element, 'keydown', {
				key: 'ArrowRight',
			});
			expect(items[1]).toEqual(document.activeElement);
		});

		it('should focus on the input when pressing the arrow right when the focus is on the first item', () => {
			component = new ClayMultiSelect({
				dataSource,
				helpText,
				selectedItems: [
					{
						label: 'Foo',
						value: 'foo',
					},
					{
						label: 'Bar',
						value: 'bar',
					},
				],
				spritemap,
			});

			const {autocomplete} = component.refs;

			triggerEvent(autocomplete.refs.input, 'keydown', {
				key: 'Backspace',
			});

			jest.runAllTimers();

			const items = autocomplete.element.querySelectorAll(
				'span[id="item-tag"]'
			);

			expect(items[1]).toEqual(document.activeElement);

			triggerEvent(autocomplete.element, 'keydown', {
				key: 'ArrowRight',
			});
			expect(autocomplete.refs.input).toEqual(document.activeElement);
		});

		it('do nothing when the focus is on the last item when press the arrow left', () => {
			component = new ClayMultiSelect({
				dataSource,
				helpText,
				selectedItems: [
					{
						label: 'Foo',
						value: 'foo',
					},
					{
						label: 'Bar',
						value: 'bar',
					},
				],
				spritemap,
			});

			const {autocomplete} = component.refs;

			triggerEvent(autocomplete.refs.input, 'keydown', {
				key: 'Backspace',
			});

			jest.runAllTimers();

			const items = autocomplete.element.querySelectorAll(
				'span[id="item-tag"]'
			);

			expect(items[1]).toEqual(document.activeElement);

			triggerEvent(autocomplete.element, 'keydown', {key: 'ArrowLeft'});
			expect(items[0]).toEqual(document.activeElement);

			triggerEvent(autocomplete.element, 'keydown', {key: 'ArrowLeft'});
			expect(items[0]).toEqual(document.activeElement);

			triggerEvent(autocomplete.element, 'keydown', {key: 'ArrowLeft'});
			expect(items[0]).toEqual(document.activeElement);
		});
	});
});
