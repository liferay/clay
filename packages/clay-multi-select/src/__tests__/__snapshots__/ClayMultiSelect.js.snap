// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayMultiSelect should render a ClayMultiSelect and clear the input values after the onblur event 1`] = `
<div class="form-group">
  <div class="input-group input-group-stacked-sm-down">
    <div class="input-group-item">
      <div class="dropdown">
        <div class=" form-control form-control-tag-group">
          <span class="autofit-row">
            <span class="autofit-col autofit-col-expand">
              <input class="form-control-inset" ref="input" type="text" value="">
            </span>
          </span>
          <div></div>
        </div>
      </div>
      <div class="form-feedback-group">
        <div class="form-text">You can use a comma to enter tags</div>
      </div>
    </div>
    <div class="input-group-item input-group-item-shrink">
      <button class="btn btn-secondary" type="button">select</button>
    </div>
  </div>
</div>
<ul class="autocomplete-dropdown-menu dropdown-menu" ref="dropdown" style="top: 0px; left: 0px;">
  <li>
    <span class="disabled dropdown-item">no-results-were-found</span>
  </li>
</ul>
`;

exports[`ClayMultiSelect should render a ClayMultiSelect and keep the focus of the input when selecting an item 1`] = `
<div class="form-group">
  <div class="input-group input-group-stacked-sm-down">
    <div class="input-group-item">
      <div class="dropdown">
        <div class=" form-control form-control-tag-group">
          <input name="selectedItems" type="hidden" value="foo">
          <span class="label label-secondary label-dismissible" id="item-tag" tabindex="0" data-tag="0">
            <span class="label-item label-item-expand">Foo</span>
            <span class="label-item label-item-after">
              <button class=" close" aria-label="close" tabindex="-1" type="button">
                <svg class="lexicon-icon lexicon-icon-times" focusable="false" role="presentation">
                  <use xlink:href="icons.svg#times"></use>
                </svg>
              </button>
            </span>
          </span>
          <input name="selectedItems" type="hidden" value="bar">
          <span class="label label-secondary label-dismissible" id="item-tag" tabindex="0" data-tag="1">
            <span class="label-item label-item-expand">Bar</span>
            <span class="label-item label-item-after">
              <button class=" close" aria-label="close" tabindex="-1" type="button">
                <svg class="lexicon-icon lexicon-icon-times" focusable="false" role="presentation">
                  <use xlink:href="icons.svg#times"></use>
                </svg>
              </button>
            </span>
          </span>
          <span class="autofit-row">
            <span class="autofit-col autofit-col-expand">
              <input class="form-control-inset" ref="input" type="text" value="">
            </span>
          </span>
          <div></div>
        </div>
      </div>
      <div class="form-feedback-group">
        <div class="form-text">You can use a comma to enter tags</div>
      </div>
    </div>
    <div class="input-group-item input-group-item-shrink">
      <button class="btn btn-secondary" type="button">select</button>
    </div>
  </div>
</div>
<ul class="autocomplete-dropdown-menu dropdown-menu" ref="dropdown">
  <li>
    <span class="disabled dropdown-item">no-results-were-found</span>
  </li>
</ul>
`;

exports[`ClayMultiSelect should render a ClayMultiSelect with classes 1`] = `
<div class="form-group my-custom-class">
  <div class="input-group input-group-stacked-sm-down">
    <div class="input-group-item">
      <div class="dropdown">
        <div class=" form-control form-control-tag-group">
          <span class="autofit-row">
            <span class="autofit-col autofit-col-expand">
              <input class="form-control-inset" ref="input" type="text" value="">
            </span>
          </span>
          <div></div>
        </div>
      </div>
      <div class="form-feedback-group">
        <div class="form-text">You can use a comma to enter tags</div>
      </div>
    </div>
    <div class="input-group-item input-group-item-shrink">
      <button class="btn btn-secondary" type="button">select</button>
    </div>
  </div>
</div>
<ul class="autocomplete-dropdown-menu dropdown-menu" ref="dropdown">
  <li>
    <span class="disabled dropdown-item">no-results-were-found</span>
  </li>
</ul>
`;

exports[`ClayMultiSelect should render a ClayMultiSelect with id 1`] = `
<div class="form-group" id="myId">
  <div class="input-group input-group-stacked-sm-down">
    <div class="input-group-item">
      <div class="dropdown">
        <div class=" form-control form-control-tag-group">
          <span class="autofit-row">
            <span class="autofit-col autofit-col-expand">
              <input class="form-control-inset" ref="input" type="text" value="">
            </span>
          </span>
          <div></div>
        </div>
      </div>
      <div class="form-feedback-group">
        <div class="form-text">You can use a comma to enter tags</div>
      </div>
    </div>
    <div class="input-group-item input-group-item-shrink">
      <button class="btn btn-secondary" type="button">select</button>
    </div>
  </div>
</div>
<ul class="autocomplete-dropdown-menu dropdown-menu" ref="dropdown">
  <li>
    <span class="disabled dropdown-item">no-results-were-found</span>
  </li>
</ul>
`;

exports[`ClayMultiSelect should render a ClayMultiSelect with input name from selected items 1`] = `
<div class="form-group">
  <div class="input-group input-group-stacked-sm-down">
    <div class="input-group-item">
      <div class="dropdown">
        <div class=" form-control form-control-tag-group">
          <span class="autofit-row">
            <span class="autofit-col autofit-col-expand">
              <input class="form-control-inset" ref="input" type="text" value="">
            </span>
          </span>
          <div></div>
        </div>
      </div>
      <div class="form-feedback-group">
        <div class="form-text">You can use a comma to enter tags</div>
      </div>
    </div>
    <div class="input-group-item input-group-item-shrink">
      <button class="btn btn-secondary" type="button">select</button>
    </div>
  </div>
</div>
<ul class="autocomplete-dropdown-menu dropdown-menu" ref="dropdown">
  <li>
    <span class="disabled dropdown-item">no-results-were-found</span>
  </li>
</ul>
`;

exports[`ClayMultiSelect should render a ClayMultiSelect with label 1`] = `
<div class="form-group">
  <label>Foo</label>
  <div class="input-group input-group-stacked-sm-down">
    <div class="input-group-item">
      <div class="dropdown">
        <div class=" form-control form-control-tag-group">
          <span class="autofit-row">
            <span class="autofit-col autofit-col-expand">
              <input class="form-control-inset" ref="input" type="text" value="">
            </span>
          </span>
          <div></div>
        </div>
      </div>
      <div class="form-feedback-group">
        <div class="form-text">You can use a comma to enter tags</div>
      </div>
    </div>
    <div class="input-group-item input-group-item-shrink">
      <button class="btn btn-secondary" type="button">select</button>
    </div>
  </div>
</div>
<ul class="autocomplete-dropdown-menu dropdown-menu" ref="dropdown">
  <li>
    <span class="disabled dropdown-item">no-results-were-found</span>
  </li>
</ul>
`;

exports[`ClayMultiSelect should render a ClayMultiSelect with placeholder 1`] = `
<div class="form-group">
  <div class="input-group input-group-stacked-sm-down">
    <div class="input-group-item">
      <div class="dropdown">
        <div class=" form-control form-control-tag-group">
          <span class="autofit-row">
            <span class="autofit-col autofit-col-expand">
              <input class="form-control-inset" ref="input" type="text" placeholder="Placeholder" value="">
            </span>
          </span>
          <div></div>
        </div>
      </div>
      <div class="form-feedback-group">
        <div class="form-text">You can use a comma to enter tags</div>
      </div>
    </div>
    <div class="input-group-item input-group-item-shrink">
      <button class="btn btn-secondary" type="button">select</button>
    </div>
  </div>
</div>
<ul class="autocomplete-dropdown-menu dropdown-menu" ref="dropdown">
  <li>
    <span class="disabled dropdown-item">no-results-were-found</span>
  </li>
</ul>
`;

exports[`ClayMultiSelect should render a ClayMultiSelect with select button hidden 1`] = `
<div class="form-group">
  <div class="input-group input-group-stacked-sm-down">
    <div class="input-group-item">
      <div class="dropdown">
        <div class=" form-control form-control-tag-group">
          <span class="autofit-row">
            <span class="autofit-col autofit-col-expand">
              <input class="form-control-inset" ref="input" type="text" value="">
            </span>
          </span>
          <div></div>
        </div>
      </div>
      <div class="form-feedback-group">
        <div class="form-text">You can use a comma to enter tags</div>
      </div>
    </div>
  </div>
</div>
<ul class="autocomplete-dropdown-menu dropdown-menu" ref="dropdown">
  <li>
    <span class="disabled dropdown-item">no-results-were-found</span>
  </li>
</ul>
`;

exports[`ClayMultiSelect should render a ClayMultiSelect with selected items 1`] = `
<div class="form-group">
  <div class="input-group input-group-stacked-sm-down">
    <div class="input-group-item">
      <div class="dropdown">
        <div class=" form-control form-control-tag-group">
          <input name="selectedItems" type="hidden" value="foo">
          <span class="label label-secondary label-dismissible" id="item-tag" tabindex="0" data-tag="0">
            <span class="label-item label-item-expand">Foo</span>
            <span class="label-item label-item-after">
              <button class=" close" aria-label="close" tabindex="-1" type="button">
                <svg class="lexicon-icon lexicon-icon-times" focusable="false" role="presentation">
                  <use xlink:href="icons.svg#times"></use>
                </svg>
              </button>
            </span>
          </span>
          <input name="selectedItems" type="hidden" value="bar">
          <span class="label label-secondary label-dismissible" id="item-tag" tabindex="0" data-tag="1">
            <span class="label-item label-item-expand">Bar</span>
            <span class="label-item label-item-after">
              <button class=" close" aria-label="close" tabindex="-1" type="button">
                <svg class="lexicon-icon lexicon-icon-times" focusable="false" role="presentation">
                  <use xlink:href="icons.svg#times"></use>
                </svg>
              </button>
            </span>
          </span>
          <span class="autofit-row">
            <span class="autofit-col autofit-col-expand">
              <input class="form-control-inset" ref="input" type="text" value="">
            </span>
          </span>
          <div></div>
        </div>
      </div>
      <div class="form-feedback-group">
        <div class="form-text">You can use a comma to enter tags</div>
      </div>
    </div>
    <div class="input-group-item input-group-item-shrink">
      <button class="btn btn-secondary" type="button">select</button>
    </div>
  </div>
</div>
<ul class="autocomplete-dropdown-menu dropdown-menu" ref="dropdown">
  <li>
    <span class="disabled dropdown-item">no-results-were-found</span>
  </li>
</ul>
`;

exports[`ClayMultiSelect should render the default markup 1`] = `
<div class="form-group">
  <div class="input-group input-group-stacked-sm-down">
    <div class="input-group-item">
      <div class="dropdown">
        <div class=" form-control form-control-tag-group">
          <span class="autofit-row">
            <span class="autofit-col autofit-col-expand">
              <input class="form-control-inset" ref="input" type="text" value="">
            </span>
          </span>
          <div></div>
        </div>
      </div>
      <div class="form-feedback-group">
        <div class="form-text">You can use a comma to enter tags</div>
      </div>
    </div>
    <div class="input-group-item input-group-item-shrink">
      <button class="btn btn-secondary" type="button">select</button>
    </div>
  </div>
</div>
<ul class="autocomplete-dropdown-menu dropdown-menu" ref="dropdown">
  <li>
    <span class="disabled dropdown-item">no-results-were-found</span>
  </li>
</ul>
`;
